File: ta4j-core/src/main/java/org/ta4j/core/criteria/ExpectedShortfallCriterion.java
/*
 * SPDX-License-Identifier: MIT
 */
package org.ta4j.core.criteria;

import java.util.Collections;
import java.util.List;
import java.util.Optional;

import org.ta4j.core.BarSeries;
import org.ta4j.core.Position;
import org.ta4j.core.TradingRecord;
import org.ta4j.core.analysis.Returns;
import org.ta4j.core.criteria.ReturnRepresentation;
import org.ta4j.core.criteria.ReturnRepresentationPolicy;
import org.ta4j.core.num.Num;
import org.ta4j.core.num.NumFactory;

/**
 * Expected Shortfall criterion, honoring the configured return representation.
 *
 * <p>
 * Measures the expected shortfall of the strategy log-return time-series.
 *
 * @see <a href=
 *      "https://en.wikipedia.org/wiki/Expected_shortfall">https://en.wikipedia.org/wiki/Expected_shortfall</a>
 *
 */
public class ExpectedShortfallCriterion extends AbstractAnalysisCriterion {

    /** Confidence level as absolute value (e.g. 0.95). */
    private final double confidence;

    private final ReturnRepresentation returnRepresentation;

    /**
     * Constructor.
     *
     * @param confidence the confidence level
     */
    public ExpectedShortfallCriterion(double confidence) {
        this(confidence, ReturnRepresentationPolicy.getDefaultRepresentation());
    }

    /**
     * Constructor.
     *
     * @param confidence           the confidence level
     * @param returnRepresentation the representation used for the result
     */
    public ExpectedShortfallCriterion(double confidence, ReturnRepresentation returnRepresentation) {
        this.confidence = confidence;
        this.returnRepresentation = returnRepresentation;
    }

    @Override
    public Optional<ReturnRepresentation> getReturnRepresentation() {
        return Optional.of(returnRepresentation);
    }

    @Override
    public Num calculate(BarSeries series, Position position) {
        if (position == null || position.getEntry() == null || position.getExit() == null) {
            return getNeutralValue(series.numFactory());
        }
        Returns returns = new Returns(series, position, ReturnRepresentation.LOG);
        return calculateES(returns, confidence, returnRepresentation);
    }

    @Override
    public Num calculate(BarSeries series, TradingRecord tradingRecord) {
        Returns returns = new Returns(series, tradingRecord, ReturnRepresentation.LOG);
        return calculateES(returns, confidence, returnRepresentation);
    }

    /**
     * Calculates the Expected Shortfall on the return series.
     *
     * @param returns    the corresponding returns
     * @param confidence the confidence level
     * @return the relative Expected Shortfall
     */
    private Num calculateES(Returns returns, double confidence, ReturnRepresentation representation) {
        // select non-NaN returns (use raw values for statistical calculations)
        List<Num> returnRates = returns.getRawValues().subList(1, returns.getSize() + 1);
        Num zero = returns.getBarSeries().numFactory().zero();
        if (returnRates.isEmpty()) {
            return getNeutralValue(returns.getBarSeries().numFactory());
        }
        Num expectedShortfall = zero;
        // F(x_var) >= alpha (=1-confidence)
        int nInBody = (int) (returns.getSize() * confidence);
        int nInTail = returns.getSize() - nInBody;

        // calculate average tail loss
        Collections.sort(returnRates);
        List<Num> tailEvents = returnRates.subList(0, nInTail);
        Num sum = zero;
        for (int i = 0; i < nInTail; i++) {
            sum = sum.plus(tailEvents.get(i));
        }
        expectedShortfall = sum.dividedBy(returns.getBarSeries().numFactory().numOf(nInTail));

        // ES is non-positive
        if (expectedShortfall.isGreaterThan(zero)) {
            expectedShortfall = zero;
        }

        // Format the final result according to the representation
        return representation.toRepresentationFromLogReturn(expectedShortfall);
    }

    /**
     * Returns the neutral value (no return) in the target representation format.
     *
     * @param numFactory the number factory
     * @return the neutral value in the target representation
     */
    private Num getNeutralValue(NumFactory numFactory) {
        if (returnRepresentation == ReturnRepresentation.MULTIPLICATIVE) {
            return numFactory.one();
        }
        // DECIMAL, PERCENTAGE, and LOG all use 0.0 as neutral
        return numFactory.zero();
    }

    /** The higher the criterion value, the better. */
    @Override
    public boolean betterThan(Num criterionValue1, Num criterionValue2) {
        return criterionValue1.isGreaterThan(criterionValue2);
    }
}

File: ta4j-core/src/main/kotlin/org/ta4j/core/criteria/ExpectedShortfallCriterion.kt
/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2017-2022 Ta4j Organization & respective
 * authors (see AUTHORS)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package org.ta4j.core.criteria

import org.ta4j.core.BarSeries
import org.ta4j.core.Position
import org.ta4j.core.TradingRecord
import org.ta4j.core.analysis.Returns
import org.ta4j.core.num.*

//import java.util.*

/**
 * Expected Shortfall criterion.
 *
 * Measures the expected shortfall of the strategy log-return time-series.
 *
 * @see [https://en.wikipedia.org/wiki/Expected_shortfall](https://en.wikipedia.org/wiki/Expected_shortfall)
 */
class ExpectedShortfallCriterion
/**
 * Constructor
 *
 * @param confidence the confidence level
 */(
    /**
     * Confidence level as absolute value (e.g. 0.95)
     */
    private val confidence: Double
) : AbstractAnalysisCriterion() {
    override fun calculate(series: BarSeries, position: Position): Num {
        if (position.entry != null && position.exit != null) {
            val returns = Returns(series, position, Returns.ReturnType.LOG)
            return calculateES(returns, confidence)
        }
        return series.numOf(0)
    }

    override fun calculate(series: BarSeries, tradingRecord: TradingRecord): Num {
        val returns = Returns(series, tradingRecord, Returns.ReturnType.LOG)
        return calculateES(returns, confidence)
    }

    /** The higher the criterion value, the better.  */
    override fun betterThan(criterionValue1: Num, criterionValue2: Num): Boolean {
        return criterionValue1.isGreaterThan(criterionValue2)
    }

    companion object {
        /**
         * Calculates the Expected Shortfall on the return series
         *
         * @param returns    the corresponding returns
         * @param confidence the confidence level
         * @return the relative Expected Shortfall
         */
        private fun calculateES(returns: Returns, confidence: Double): Num {
            // select non-NaN returns
            val returnRates = returns.values.subList(1, returns.size + 1)
            val zero = returns.numOf(0)
            var expectedShortfall = zero
            if (returnRates.isNotEmpty()) {
                // F(x_var) >= alpha (=1-confidence)
                val nInBody = (returns.size * confidence).toInt()
                val nInTail = returns.size - nInBody

                // calculate average tail loss
                returnRates.sort()
                val tailEvents = returnRates.subList(0, nInTail)
                var sum = zero
                for (i in 0 until nInTail) {
                    sum = sum.plus(tailEvents[i])
                }
                expectedShortfall = sum.div(returns.numOf(nInTail))

                // ES is non-positive
                if (expectedShortfall.isGreaterThan(zero)) {
                    expectedShortfall = zero
                }
            }
            return expectedShortfall
        }
    }
}